File: golang/day08/main.go
package main

import (
	"bufio"
	"os"
	"regexp"
	"strings"
)

type Direction int

const (
	R = iota
	L
)

type Node struct {
	Left  *Node
	Right *Node
}

func gcd(a, b int) int {
	for b != 0 {
		a, b = b, a%b
	}
	return a
}

func lcm(a, b int) int {
	return a * b / gcd(a, b)
}

func loadData(filename string) ([]Direction, map[string]map[Direction]string, error) {
	data, err := os.Open(filename)
	if err != nil {
		return nil, nil, err
	}
	defer func(data *os.File) {
		err := data.Close()
		if err != nil {
			panic(err)
		}
	}(data)
	scanner := bufio.NewScanner(data)
	scanner.Scan()
	directions := make([]Direction, 0)
	for _, d := range scanner.Text() {
		switch d {
		case 'R':
			directions = append(directions, R)
		case 'L':
			directions = append(directions, L)
		}
	}
	scanner.Scan()
	network := make(map[string]map[Direction]string)
	matcher, err := regexp.Compile(`\((\w{3}), (\w{3})\)`)
	if err != nil {
		return nil, nil, err
	}
	for scanner.Scan() {
		split := strings.Split(scanner.Text(), " = ")
		destinations := matcher.FindStringSubmatch(split[1])
		network[split[0]] = make(map[Direction]string)
		network[split[0]][L] = destinations[1]
		network[split[0]][R] = destinations[2]
	}
	return directions, network, nil
}

func followPath(start string, directions []Direction, network map[string]map[Direction]string, checkFn func(string) bool) int {
	currentNode := start
	steps := 0
	currentIndex := 0
	for checkFn(currentNode) {
		currentNode = network[currentNode][directions[currentIndex]]
		steps++
		currentIndex = (currentIndex + 1) % len(directions)
	}
	return steps
}

func solutionPart1() {
	directions, network, err := loadData("input.txt")
	if err != nil {
		panic(err)
	}
	steps := followPath("AAA", directions, network, func(node string) bool { return node != "ZZZ" })
	println(steps)
}

func solutionPart2() {
	directions, network, err := loadData("input.txt")
	if err != nil {
		panic(err)
	}
	cycles := make([]int, 0)
	for key := range network {
		if strings.HasSuffix(key, "A") {
			cycles = append(cycles, followPath(key, directions, network, func(node string) bool { return !strings.HasSuffix(node, "Z") }))
		}
	}
	currentLcm := 1
	for _, cycle := range cycles {
		currentLcm = lcm(currentLcm, cycle)
	}
	println(currentLcm)
}

func main() {
	solutionPart1()
	solutionPart2()
}

File: Rust/day08/src/main.rs
use std::collections::HashMap;
use regex::Regex;
use num_integer::lcm;

#[derive(Debug)]
enum Direction {
    R,
    L,
}

fn load_data(filename: &str) -> (Vec<Direction>, HashMap<String, (String, String)>) {
    let binding = std::fs::read_to_string(filename).unwrap();
    let mut data = binding.lines();
    let mut instructions = Vec::new();
    for direction in data.next().unwrap().chars() {
        if direction == 'R' {
            instructions.push(Direction::R);
        } else {
            instructions.push(Direction::L);
        }
    }
    let mut network = HashMap::new();
    data.next();
    let re = Regex::new(r"\((\w{3}), (\w{3})\)").unwrap();
    for line in data {
        let mut line = line.split(" = ");
        let address = line.next().unwrap().to_string();
        let value = line.next().unwrap().to_string();
        match re.captures(&value) {
            Some(caps) => {
                let value = (caps[1].to_string(), caps[2].to_string());
                network.insert(address, value);
            }
            None => {
                let value = (value, "".to_string());
                network.insert(address, value);
            }
        }
    }
    return (instructions, network);
}

fn follow_path(starts: &str,
               instructions: &Vec<Direction>,
               network: &HashMap<String, (String, String)>,
               terminate_fn: &dyn Fn(&str) -> bool) -> i32 {
    let mut current = starts;
    let mut current_instruction_index = 0;
    let mut steps = 0;
    while terminate_fn(current) {
        steps += 1;
        match instructions[current_instruction_index] {
            Direction::L => {
                let (next, _) = network.get(current).unwrap();
                // println!("{} -> {}", current, next);
                current = next;
                current_instruction_index += 1;
            }
            Direction::R => {
                let (_, next) = network.get(current).unwrap();
                // println!("{} -> {}", current, next);
                current = next;
                current_instruction_index += 1;
            }
        }
        if current_instruction_index == instructions.len() {
            current_instruction_index = 0;
        }
    }
    return steps;
}
fn solution_part_1() {
    let (instruction, network) = load_data("input.txt");
    let steps = follow_path("AAA", &instruction, &network, &|current| current != "ZZZ");
    println!("Steps: {}", steps);
}

fn solution_part_2() {
    let (instruction, network) = load_data("input.txt");
    // find nodes ending in "A"
    let mut cycles: Vec<i64> = Vec::new();
    for node in network.keys() {
        if node.ends_with("A") {
            cycles.push(follow_path(node, &instruction, &network, &|current| !current.ends_with("Z")) as i64);
        }
    }
    let cycles = cycles.iter().fold(1, |acc, x| lcm(acc, *x));
    println!("Cycles: {:?}", cycles);
}
fn main() {
    solution_part_1();
    solution_part_2()
}
